### PR TITLE
Parallelized cache, text and JSON parsing

### PR DESCRIPTION
**Pull request by nishantkr18**
### Without `-k, -c, --passes`:
- [x] `--num_parse_threads=1`
- [x] `--num_parse_threads=10`

### With `-k, -c`, but without `--passes`:
- [x] `--num_parse_threads=1`
- [x] `--num_parse_threads=10` 

### Without `-k`, but with `-c, --passes`:
- [x] `--passes=10 --num_parse_threads=1`
- [x] `--passes=10 --num_parse_threads=10` 
- [x] `--passes=1 --num_parse_threads=10` 

## JSON input type:
- [x] `--json`

### CODE REVIEW

File: vowpalwabbit/example.h [hunk 1]
```diff
@@ -158,11 +158,35 @@ struct example : public example_predict  // core example datatype.
...
   bool in_use = true;
 
-  // This object is only used because atomic signaling is not available in C++11
-  example_lock ex_lock;
+  // // This object is only used because atomic signaling is not available in C++11
+  // example_lock ex_lock;
 };
 VW_WARNING_STATE_POP
 
+struct example_vector  // core example datatype.
```
Comment (peterychang) [inline vowpalwabbit/example.h]: There are 2 ways we can handle this.

1. We can use a  custom object like you have here. This will mean that we need to remove all the parsing_done stuff out of example and move it into this object. We would also want to use the definition of multi_ex instead of a v_array for the example vector so that we can send it down the learner stack directly
https://github.com/VowpalWabbit/vowpal_wabbit/blob/f53a8d96a04a2c86ad3eb1d1ba27587182dda435/vowpalwabbit/reductions_fwd.h#L13
2. We can use the knowledge that all the examples in one vector will be done sequentially in a single thread. In this case, we simply keep the FIRST example (index 0 in the vector) as `done_parsing == false` until every example is parsed. Then the learner thread (in get_example) can just check `example_vector[0].done_parsing`.

The first method is truer to the actual state, but the second is a lot easier to implement since it requires a lot less testing and validation. Personally I would recommend the second approach since it's a lot less invasive and the code will be a lot simpler for now

File: vowpalwabbit/io_buf.cc [hunk 2]
```diff
@@ -90,13 +90,13 @@ void io_buf::buf_write(char*& pointer, size_t n)
   }
   else  // Time to dump the file
   {
-    if (head != _buffer._begin)
-      flush();
-    else  // Array is short, so increase size.
-    {
+    // if (head != _buffer._begin)
```
Comment (peterychang) [inline vowpalwabbit/io_buf.cc:93]: the io_buf object is used in several places. Don't make changes to existing functions without investigating the exactly how it will affect other workflows. If you really need to, make a new function thats specialized to your need (and name it appropriately so we know its a candidate for refactoring work)

File: vowpalwabbit/vw.h [hunk 3]
```diff
@@ -118,7 +118,8 @@ void parse_example_label(vw& all, example& ec, std::string label);
 void setup_examples(vw& all, v_array<example*>& examples);
 void setup_example(vw& all, example* ae);
 example* new_unused_example(vw& all);
-example* get_example(parser* pf);
+example_vector* get_example(vw& pf);
+void work_on_example(vw& pf, example* ex);
```
Comment (peterychang) [inline vowpalwabbit/vw.h:122]: You shouldn't add anything to this header file. `vw.h` is specifically meant to be our public-facing API and none of the work you're doing should be exposed to outside callers

File: vowpalwabbit/io_to_queue.h [hunk 4]
```diff
@@ -69,18 +52,32 @@ inline bool read_input_file_ascii(vw& all, char *&line) {
 }
  
 inline bool read_input_file_binary(vw& all, char *&line) {
+
+  size_t num_chars_initial;
```
Comment (peterychang) [inline vowpalwabbit/io_to_queue.h]: nit: rename this to something a little more descriptive. something like `example_length_bytes`

File: vowpalwabbit/io_to_queue.h [hunk 5]
```diff
@@ -69,18 +52,32 @@ inline bool read_input_file_ascii(vw& all, char *&line) {
...
+  all.example_parser->input->set(line);
+  (*input).buf_read(line, num_chars_initial);
  
- size_t total_num_input_bytes = 0;
- 
- //read_all_data returns the number of bytes successfully read from the input files
- total_num_input_bytes = all.example_parser->input->read_all_data(line, total_num_input_bytes);
- 
- bool finish = true;
+ bool finish = false;
```
Comment (peterychang) [inline vowpalwabbit/io_to_queue.h]: nit: `finish` never changes, so you can just get rid of this and return false at the end

File: vowpalwabbit/io_to_queue.h [hunk 6]
```diff
@@ -69,18 +52,32 @@ inline bool read_input_file_ascii(vw& all, char *&line) {
...
  
 inline bool read_input_file_binary(vw& all, char *&line) {
+
+  size_t num_chars_initial;
+  io_buf* input = all.example_parser->input.get();
+
+  // Return true if no examples found in file.
+  if((*input).buf_read(line, sizeof(num_chars_initial)) < sizeof(num_chars_initial)) {
+    // The end of io_lines should contain a null pointer.
+    std::vector<char> *byte_array = new std::vector<char>();
```
Comment (peterychang) [inline vowpalwabbit/io_to_queue.h]: We'll eventually want to use an object pool for these vectors too. You can leave it alone for now and we'll get to it when we start optimizing for speed toward the end of the project

File: vowpalwabbit/io_to_queue.h [hunk 7]
```diff
@@ -69,18 +52,32 @@ inline bool read_input_file_ascii(vw& all, char *&line) {
...
+    all.example_parser->io_lines.push(std::move(byte_array));
+    return true;
+  }
+  // read the example size.
+  num_chars_initial = *(size_t*)line;
+  line += sizeof(num_chars_initial);
+  all.example_parser->input->set(line);
+  (*input).buf_read(line, num_chars_initial);
```
Comment (peterychang) [inline vowpalwabbit/io_to_queue.h]: Check the return value and do an assert
```
auto total_read = input->buf_read(line, expected_size);
assert(total_read = expected_size);
```

If these are not the same, then we're in an error condition that we might not be able to recover from

File: vowpalwabbit/learner.cc [hunk 8]
```diff
@@ -215,12 +220,32 @@ class multi_example_handler
 class ready_examples_queue
 {
 public:
-  ready_examples_queue(vw& master) : _master(master) {}
+  ready_examples_queue(vw& master) : _master(master), _index(0) {}
+
+  example* pop() {
```
Comment (peterychang) [inline vowpalwabbit/learner.cc]: You should be returning `vector<example*>*` here

File: vowpalwabbit/parse_dispatch_loop.h [hunk 9]
```diff
@@ -12,66 +12,66 @@
...
-
-      if (!all.do_reset_source && example_number != all.pass_length && all.max_examples > example_number &&
-          all.example_parser->reader(&all, examples, words_localcpy, parse_name_localcpy) > 0)
-      {
+      example_vector* examples = &VW::get_unused_example_vector(&all);
+      (*examples).push_back(example_ptr);
 
-        VW::setup_examples(all, examples);
-        example_number += examples.size();
+      bool is_null_pointer = false;
```
Comment (peterychang) [inline vowpalwabbit/parse_dispatch_loop.h]: whats the point of this variable? You can just check `io_lines_next_item == nullptr`

File: vowpalwabbit/parse_dispatch_loop.h [hunk 10]
```diff
@@ -12,66 +12,66 @@
...
 
-        dispatch(all, examples);
+      // This logic should be outside the reader, as:
+      // 1. it is common to all types of parsers.
+      // 2. the type of parser might change when the thread is waiting for a pop of io_lines.
+      std::vector<char> *io_lines_next_item;
```
Comment (peterychang) [inline vowpalwabbit/parse_dispatch_loop.h]: nit: set this to nullptr

File: vowpalwabbit/parse_example.cc [hunk 11]
```diff
@@ -21,22 +21,22 @@
 
 namespace logger = VW::io::logger;
 
-size_t read_features(vw *all, std::vector<char>& line, size_t&, v_array<example*>& examples)
+int read_features(vw *all, std::vector<char>& line, size_t&, v_array<example*>& examples)
```
Comment (peterychang) [inline vowpalwabbit/parse_example.cc]: just delete this function

File: vowpalwabbit/parse_example.cc [hunk 12]
```diff
@@ -63,17 +63,16 @@ size_t strip_features_string(char*& line, size_t num_chars_init){
...
   size_t num_chars;
-  size_t num_chars_initial;
+  size_t num_chars_initial = (*io_lines_next_item).size();
 
   // a line is popped off of the io queue in read_features
-  num_chars_initial = read_features(all, line, num_chars, examples);
+  // num_chars_initial = read_features(all, line, num_chars, examples);
 
-  char *stripped_line = std::move(line.data());
+  char *stripped_line = std::move((*io_lines_next_item).data());
```
Comment (peterychang) [inline vowpalwabbit/parse_example.cc]: don't do a move here, its actually wrong. Practically it doesn't make a difference, but in theory the source can be changed to an ambiguous value, which we don't want

File: vowpalwabbit/parse_example.h [hunk 13]
```diff
@@ -19,13 +19,14 @@ void substring_to_example(vw* all, example* ae, VW::string_view& example, std::v
...
 void read_line(vw& all, example* ex, char* line);  // read example from the line.
 void read_lines(vw* all, const char* line, size_t len,
     v_array<example*>& examples);  // read examples from the new line separated strings.
 
 }  // namespace VW
 
-int read_features_string(vw* all, v_array<example*>& examples, std::vector<VW::string_view>& words, std::vector<VW::string_view>& parse_name);
-size_t read_features(vw *all, std::vector<char>& line, size_t& num_chars, v_array<example*>& examples);
+int read_features_string(vw* all, v_array<example*>& examples, std::vector<VW::string_view>& words, std::vector<VW::string_view>& parse_name, std::vector<char> *io_lines_next_item);
+int read_features(vw *all, std::vector<char>& line, size_t& num_chars, v_array<example*>& examples);
```
Comment (peterychang) [inline vowpalwabbit/parse_example.h]: delete this function

File: vowpalwabbit/parser.cc [hunk 14]
```diff
@@ -623,7 +623,20 @@ void enable_sources(vw& all, bool quiet, size_t passes, input_options& input_opt
...
+  // p.can_end_pass_parser.notify_all();
   p.done = true;
+
+  // lock_done() can be called by the learner thread when the examples have reached the maximum pass allowed.
+  // So, in such a condition, an un-timely end of the program is triggered.
+  // That is why, we release the io_thread in such un-timely situations.
+  p.done_with_io.store(true);
+  p.can_end_pass.notify_one();
+
+  p.io_lines.set_done();
```
Comment (peterychang) [inline vowpalwabbit/parser.cc:639]: make sure these operations are done in an order that makes sense programmatically. A few questions you should be asking yourself
1. When this function is called, where can each thread be in execution?
2. What race conditions will be present with different execution orders?
3. Do any of the race conditions affect the end result?

Also, be aware that threads can wake up at any time, even without being notified. This is the reason they typically live inside a conditional loop

File: vowpalwabbit/parser.cc [hunk 15]
```diff
@@ -668,9 +681,20 @@ example& get_unused_example(vw* all)
   return *ex;
 }
 
+example_vector& get_unused_example_vector(vw* all)
+{
+  parser* p = all->example_parser;
+  auto ex_vector = p->example_vector_pool.get_object();
+  VW_WARNING_STATE_PUSH
```
Comment (peterychang) [inline vowpalwabbit/parser.cc]: you're not using anything deprecated? Get rid of these lines

File: vowpalwabbit/vwdll.cpp [hunk 16]
```diff
@@ -149,7 +149,7 @@ VW_DLL_PUBLIC void VW_CALLING_CONV VW_EndParser(VW_HANDLE handle)
 VW_DLL_PUBLIC VW_EXAMPLE VW_CALLING_CONV VW_GetExample(VW_HANDLE handle)
 { vw * pointer = static_cast<vw*>(handle);
   parser * parser_pointer = static_cast<parser *>(pointer->example_parser);
-  return static_cast<VW_EXAMPLE>(VW::get_example(parser_pointer));
+  return static_cast<VW_EXAMPLE>(VW::get_example(*pointer));
```
Comment (peterychang) [inline vowpalwabbit/vwdll.cpp]: don't touch anything in this file. This is our DLL binding layer and any changes here can be dangerous

**Review by peterychang — COMMENTED** [review]
A few things. First off, you can ignore anything marked as 'nit' if you want. They're minor things (usually stylistic) that I don't particularly care about either way.
That said: nit: do `var->size()` instead of `(*var).size()`.

If you want to use `vector<example*>` for the json parser, then we need to treat it as a first class citizen. Basically the workflow splits between these two handlers
single example: https://github.com/VowpalWabbit/vowpal_wabbit/blob/f53a8d96a04a2c86ad3eb1d1ba27587182dda435/vowpalwabbit/learner.cc#L140
multi example: https://github.com/VowpalWabbit/vowpal_wabbit/blob/f53a8d96a04a2c86ad3eb1d1ba27587182dda435/vowpalwabbit/learner.cc#L162

You can see they handle the `on_example` function differently. With vectors, you should be sending `vector<example*>` into these functions. Then, instead of the multi_ex handler aggregating several examples together, the single example handler should be pulling the (hopefully one) example out of the vector.
This vector should be the same one that was pool allocated in the parser (or at worst a vector where the contents were `move`d from the pool allocated vector)

In short, if we use a vector, then every function from the parser to the learner handlers need to be modified to take in vectors

This is the main reason I mentioned using lists in our last meeting. Using a vector will probably be the solution we use in production, but it is much more complex and time consuming to implement.

File: vowpalwabbit/example.h (hunk 1 above)
Comment (nishantkr18) [reply vowpalwabbit/example.h]: With the current changes, if I were to use the second method, I should just move the `example_lock` object from the `example_vector`, back to the `example` struct.(and then replace the use of `example_vector` with `multi_ex` or `v_array<examples*>`) right?

I think we will still require the use of something like `get_unused_example_vector()` were we fetch an example vector from the object pool. 

Also, I can see that the parser utilizes `v_array<example *>` everywhere. Yet we also have `multi_ex`. What are their usecases, how are they different and which one should we use here?

File: vowpalwabbit/io_buf.cc (hunk 2 above)
Comment (nishantkr18) [reply vowpalwabbit/io_buf.cc:93]: Oh yes, I'd checked where the function `buf_write` was used in the entire vw library, and I realized that I wanted to change the function's behaviour for each of those places (basically, anywhere we write cache to the buffer). 

And btw, all of these calls are made from the cache writing logic in the `setup_example` function.

File: vowpalwabbit/vwdll.cpp (hunk 16 above)
Comment (nishantkr18) [reply vowpalwabbit/vwdll.cpp]: If I were to change the parameter types of the function `get_example` say, then without making the corresponding change here, the compiler gives an error. So, I should avoid changing the parameter types of functions used here?

**Comment by nishantkr18** [comment]
> You can see they handle the `on_example` function differently. With vectors, you should be sending `vector<example*>` into these functions. Then, instead of the multi_ex handler aggregating several examples together, the single example handler should be pulling the (hopefully one) example out of the vector.

If we directly send `vector<example*>`  into the handler, we need to make sure we perform the setup_examples and the special conditional "end of pass" logic (dealing with the pass counting and containing `reset_source`) before hand. I say this because the handler does not have the vw object reference `all` passed to it.  I thought to make it work, it will require a lot of unwanted changes, so I didnt touch the handler code. What do you think?

File: vowpalwabbit/example.h (hunk 1 above)
Comment (peterychang) [reply vowpalwabbit/example.h]: the v_array usage in the parser is mostly a legacy thing. The multi_ex definition used to be `v_array<example*>` but it was converted to `vector<example*>` sometime in the last year or so. However, since the parser and learner's usage of them are completely separate, we never went and refactored the parser.

File: vowpalwabbit/io_buf.cc (hunk 2 above)
Comment (peterychang) [reply vowpalwabbit/io_buf.cc:93]: Did you also check the language bindings? I know at least one of them made use of the io_buf

File: vowpalwabbit/vwdll.cpp (hunk 16 above)
Comment (peterychang) [reply vowpalwabbit/vwdll.cpp]: In general, changing any code in a dynamic library's binding layer can be extremely problematic. There are things you can safely change, but it generally requires extreme care. Since this is in a .cpp file, its _probably_ safe, but I currently don't have the time to go in and analyze the implications of any changes made here. So in general, its best to simply avoid changing any of the code explicitly exposed in the APIs.

If you feel its absolutely necessary, feel free to create new functions (add their prototypes and definitions in their own file if you need to). We can go back and fix it up later once a full evaluation's been done

**Comment by peterychang** [comment]
>  I say this because the handler does not have the vw object reference all passed to it.

Its in the context object
https://github.com/VowpalWabbit/vowpal_wabbit/blob/f53a8d96a04a2c86ad3eb1d1ba27587182dda435/vowpalwabbit/learner.cc#L158
https://github.com/VowpalWabbit/vowpal_wabbit/blob/f53a8d96a04a2c86ad3eb1d1ba27587182dda435/vowpalwabbit/learner.cc#L102

File: vowpalwabbit/io_buf.cc (hunk 2 above)
Comment (nishantkr18) [reply vowpalwabbit/io_buf.cc:93]: Yeah, but `buf_write` is not used in any of them.

File: test/tools/parser_throughput/main.cc [hunk 17]
```diff
@@ -181,7 +181,7 @@ int main(int argc, char** argv)
       }
       // TODO - finish_example should support a v_array as input.
       VW::finish_example(*vw, result);
-      examples.delete_v();
+      examples.clear();
```
Comment (peterychang) [inline test/tools/parser_throughput/main.cc]: nit: unnecessary. The vector's destructor will handle this

File: vowpalwabbit/vw.h [hunk 18]
```diff
@@ -115,10 +115,11 @@ void dealloc_example(void (*delete_label)(polylabel*), example& ec, void (*delet
 void dealloc_examples(example* example_ptr, size_t count);
 
 void parse_example_label(vw& all, example& ec, std::string label);
-void setup_examples(vw& all, v_array<example*>& examples);
+void setup_examples(vw& all, std::vector<example*>& examples);
```
Comment (peterychang) [inline vowpalwabbit/vw.h:118]: since we intend for this to eventually go into master, we need to be extremely careful when we're modifying this file. These vector changes breaks the API, which is probably not ok. The correct way to handle this would be to create new entry points, keep (but deprecate) the old interfaces, then write a translation layer between the two

File: vowpalwabbit/io_to_queue.h [hunk 19]
```diff
@@ -69,22 +52,35 @@ inline bool read_input_file_ascii(vw& all, char *&line) {
...
- 
+
+  size_t example_length_bytes;
+  io_buf* input = all.example_parser->input.get();
+
+  // Return true if no examples found in file.
+  if((*input).buf_read(line, sizeof(example_length_bytes)) < sizeof(example_length_bytes)) {
+    // The end of io_lines should contain a null pointer.
+    std::vector<char> *byte_array = new std::vector<char>();
+    all.example_parser->io_lines.push(std::move(byte_array));
```
Comment (peterychang) [inline vowpalwabbit/io_to_queue.h]: nit: the move here is unnecessary. You're attempting to move a pointer, which is just a copy

File: vowpalwabbit/learner.cc [hunk 20]
```diff
@@ -142,16 +147,23 @@ class single_example_handler
...
   {
-    if (ec->indices.size() > 1)  // 1+ nonconstant feature. (most common case first)
-      _context.template process<example, learn_ex>(*ec);
-    else if (ec->end_pass)
-      _context.template process<example, end_pass>(*ec);
-    else if (is_save_cmd(ec))
-      _context.template process<example, save>(*ec);
-    else
-      _context.template process<example, learn_ex>(*ec);
+    for (example* ec : *ev)
```
Comment (peterychang) [inline vowpalwabbit/learner.cc]: nit: technically this is unnecessary since the vector is guaranteed to have exactly 1 element at this point. We can leave it alone for now, but directly indexing the first value might be slightly faster (depends on how well the compiler optimizes this stuff)

File: vowpalwabbit/learner.cc [hunk 21]
```diff
@@ -196,13 +208,20 @@ class multi_example_handler
...
+  void on_example(std::vector<example*>* ev)
   {
-    if (try_complete_multi_ex(ec))
+    for (example* ec : *ev)
     {
-      _context.template process<multi_ex, learn_multi_ex>(ec_seq);
-      ec_seq.clear();
+      work_on_example(_context.get_master(), ec);
+      
+      if (try_complete_multi_ex(ec))
```
Comment (peterychang) [inline vowpalwabbit/learner.cc]: you'll need to rewrite the `try_complete_multi_ex()` function. Its taking an example, and if its part of a multi-example then its compiling them into another vector that will get sent down the learning stack. Thats inefficient, we should just send the vector we already have.

Basically, what we should be doing is:
```
if (ev[0]->indices.size() > 1)
  complete_multi_ex(ev)
  process<learn_multi_ex>()
else if (ev[0]->end_pass)
  process<end_pass>()
else if (is_save_cmd(ev[0]))
  process<save>()
else
  complete_multi_ex(ev)
  process<learn_multi_ex>
```

complete_multi_ex would need a bit of work in this case

File: vowpalwabbit/learner.cc [hunk 22]
```diff
@@ -280,12 +299,12 @@ void generic_driver_onethread(vw& all)
 {
   single_instance_context context(all);
   handler_type handler(context);
-  auto multi_ex_fptr = [&handler](vw& all, v_array<example*> examples) {
+  auto multi_ex_fptr = [&handler](vw& all, std::vector<example*> examples) {
     all.example_parser->end_parsed_examples += examples.size();  // divergence: lock & signal
-    custom_examples_queue examples_queue(examples);
-    process_examples(examples_queue, handler);
+    // custom_examples_queue examples_queue(examples);
```
Comment (peterychang) [inline vowpalwabbit/learner.cc]: Why is this commented out?

File: vowpalwabbit/parse_dispatch_loop.h [hunk 23]
```diff
@@ -12,66 +12,63 @@
...
 
-        reset_source(all, all.num_bits);
 
-        // to call reset source in io thread
-        all.example_parser->done_with_io.store(true);
-        all.example_parser->can_end_pass.notify_one();
-
-        all.do_reset_source = false;
-        all.passes_complete++;
+      if(io_lines_next_item == nullptr) {
```
Comment (peterychang) [inline vowpalwabbit/parse_dispatch_loop.h]: add a comment mentioning this can only happen after all the parsing is finished (including all the passes)

File: vowpalwabbit/parser.cc [hunk 24]
```diff
@@ -925,35 +963,75 @@ void finish_example(vw& all, example& ec)
...
+  all.example_parser->example_vector_pool.return_object(&ev);
+}
 }  // namespace VW
 
-void thread_dispatch(vw& all, const v_array<example*>& examples)
+void thread_dispatch(vw& all, const std::vector<example*>& examples)
 {
-  all.example_parser->end_parsed_examples += examples.size();
-  notify_examples_cv(examples[0]);
+  for (auto example : examples) notify_examples_cv(example);
```
Comment (peterychang) [inline vowpalwabbit/parser.cc]: Instead of checking and notifying every example in the vector, you can just check and notify the first example to signify that the entire vector is complete. This'll save a lot of expensive calls.

In fact, you do only check the first example below

File: vowpalwabbit/parser.cc [hunk 25]
```diff
@@ -925,35 +963,75 @@ void finish_example(vw& all, example& ec)
...
+  else{
+    reset_source(all, all.num_bits);
+    // all.do_reset_source = false;
+    all.passes_complete++;
 
+    // setup an end_pass example
+    all.example_parser->lbl_parser.default_label(&ex->l);
+    all.example_parser->in_pass_counter = 0;
+
+    // if (all.passes_complete == all.numpasses && example_number == all.pass_length)
```
Comment (peterychang) [inline vowpalwabbit/parser.cc]: why are these lines commented out?

File: vowpalwabbit/learner.cc (hunk 21 above)
Comment (nishantkr18) [reply vowpalwabbit/learner.cc]: initially, we created an example sequence by pushing back single examples into the sequence.

Now, we should create a sequence by pushing back all elements of each of the example vectors received by `on_example()` into the sequence.

Does that sound right?

File: vowpalwabbit/learner.cc (hunk 22 above)
Comment (nishantkr18) [reply vowpalwabbit/learner.cc]: Was facing an issue, is resolved now.

File: vowpalwabbit/parser.cc (hunk 25 above)
Comment (nishantkr18) [reply vowpalwabbit/parser.cc]: Taken care of by adding a global example_number :+1:

File: vowpalwabbit/io_to_queue.h [hunk 26]
```diff
@@ -69,22 +52,34 @@ inline bool read_input_file_ascii(vw& all, char *&line) {
...
- //read_all_data returns the number of bytes successfully read from the input files
- total_num_input_bytes = all.example_parser->input->read_all_data(line, total_num_input_bytes);
- 
- bool finish = true;
- 
+
+  size_t example_length_bytes;
+  io_buf* input = all.example_parser->input.get();
+
+  // Return true if no examples found in file.
```
Comment (jackgerrits) [inline vowpalwabbit/io_to_queue.h]: True if n examples found seems counterintuitive to me

Comment (nishantkr18) [reply vowpalwabbit/io_to_queue.h]: Just modified the comment. Maybe it makes more sense now.

File: vowpalwabbit/parse_dispatch_loop.h (hunk 23 above)
Comment (nishantkr18) [reply vowpalwabbit/parse_dispatch_loop.h]: Doesn't this happen at the end of each pass?
